pkg/reconciler/workspace/controller.go

```diff
@@ -0,0 +1,385 @@
+	if err != nil {
+		return err
+	}
+	previous := obj.DeepCopy()
+
+	if err := c.reconcile(ctx, obj); err != nil {
+		return err
+	}
+
+	// If the object being reconciled changed as a result, update it.
+	if !equality.Semantic.DeepEqual(previous.Spec, obj.Spec) {
+		_, uerr := c.kcpClient.Cluster(clusterName).TenancyV1alpha1().Workspaces().Update(ctx, obj, metav1.UpdateOptions{})
+		return uerr
+	}
+	if !equality.Semantic.DeepEqual(previous.Status, obj.Status) {
+		_, uerr := c.kcpClient.Cluster(clusterName).TenancyV1alpha1().Workspaces().UpdateStatus(ctx, obj, metav1.UpdateOptions{})
+		return uerr
+	}
+
+	return nil
+}
+
+func (c *Controller) reconcile(ctx context.Context, obj *tenancyv1alpha1.Workspace) error {
+	if currentShard := obj.Status.Location.Current; currentShard != "" {
+		// make sure current shard still exists
+		shards, err := c.workspaceShardLister.List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		var found bool
+		for _, shard := range shards {
+			if shard.Name == currentShard {
+				found = true
+				break
+			}
+		}
+		if !found {
+			klog.Infof("removing spec.targetShard for workspace %q", obj.Name)
+			obj.Spec.TargetShard = ""
+		}
+	}
+	if obj.Spec.TargetShard == "" {
+		// find a shard for this workspace
+		shards, err := c.workspaceShardLister.List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		if len(shards) != 0 {
+			target := shards[rand.Intn(len(shards))].Name
+			obj.Spec.TargetShard = target
+			obj.Status.Location.Target = target
+			klog.Infof("moving workspace %q to %q", obj.Name, target)
+		}
+	}
+	if obj.Spec.TargetShard != obj.Status.Location.Current {
+		klog.Infof("updating workspace %q status", obj.Name)
+		obj.Status.Location.Current = obj.Spec.TargetShard
+		obj.Status.Location.Target = ""
+	}
+	if obj.Status.Location.Current == "" {
+		obj.Status.Phase = tenancyv1alpha1.WorkspacePhaseInitializing
+		if !IsWorkspaceUnschedulable(obj) {
+			klog.Infof("marking workspace %q unschedulable", obj.Name)
+			SetWorkspaceCondition(obj, tenancyv1alpha1.WorkspaceCondition{
+				Type:               tenancyv1alpha1.WorkspaceScheduled,
+				Status:             metav1.ConditionFalse,
+				LastProbeTime:      metav1.Time{Time: time.Now()},
+				LastTransitionTime: metav1.Time{Time: time.Now()},
+				Reason:             tenancyv1alpha1.WorkspaceReasonUnschedulable,
+				Message:            "No shards are available to schedule Workspaces to.",
+			})
+		}
+	} else {
+		obj.Status.Phase = tenancyv1alpha1.WorkspacePhaseAtive
+		if IsWorkspaceUnschedulable(obj) {
+			klog.Infof("marking workspace %q scheduled", obj.Name)
+			SetWorkspaceCondition(obj, tenancyv1alpha1.WorkspaceCondition{
+				Type:               tenancyv1alpha1.WorkspaceScheduled,
+				Status:             metav1.ConditionTrue,
+				LastProbeTime:      metav1.Time{Time: time.Now()},
+				LastTransitionTime: metav1.Time{Time: time.Now()},
```

Comment:
this is bad style. You logic might avoid hotlooping, but better have a helper func that sets the condition, comparing the old to the new status and only then set LastTransitionTime.

Reply:
https://github.com/kcp-dev/kcp/issues/246